[Flaky Test] pull-cluster-api-provider-gardener-unit:

<!-- Please only use this template for submitting reports about flaky tests or jobs (pass or fail with no underlying change in code) in Gardener CI -->

**How to categorize this issue?**
<!--
Please select area, kind, and priority for this issue. This helps the community categorizing it.
Replace below TODOs or exchange the existing identifiers with those that fit best in your opinion.
If multiple identifiers make sense you can also state the commands multiple times, e.g.
  /area control-plane
  /area auto-scaling
  ...

"/area" identifiers:     audit-logging|auto-scaling|backup|certification|control-plane-migration|control-plane|cost|delivery|dev-productivity|disaster-recovery|documentation|high-availability|logging|metering|monitoring|networking|open-source|ops-productivity|os|performance|quality|robustness|scalability|security|storage|testing|usability|user-management
"/kind" identifiers:     api-change|bug|cleanup|discussion|enhancement|epic|impediment|poc|post-mortem|question|regression|task|technical-debt|test
-->
/area testing
/kind flake

**Which test(s)/suite(s) are flaking**:
`github.com/gardener/cluster-api-provider-gardener/internal/controller/controlplane`

**CI link**:
- https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_cluster-api-provider-gardener/31/pull-cluster-api-provider-gardener-unit/1945069437379940352
- https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_cluster-api-provider-gardener/30/pull-cluster-api-provider-gardener-unit/1945059419561660416

**Reason for failure**:
```
panic: test timed out after 10m0s
	running tests:
		TestControllers (10m0s)
[...]
FAIL	github.com/gardener/cluster-api-provider-gardener/internal/controller/controlplane	600.063s
```

**Anything else we need to know**:

--- a/Makefile
+++ b/Makefile
@@ -98,8 +98,8 @@ vet: ## Run go vet against code.
 	go vet ./...
 
 .PHONY: test
-test: setup-envtest ## Run tests.
-	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_BRANCH_NAME) --bin-dir $(LOCALBIN) -p path)" go test $$(go list ./... | grep -v /e2e) -coverprofile cover.out
+test: $(REPORT_COLLECTOR) $(SETUP_ENVTEST) ## Run tests.
+	@bash $(GARDENER_HACK_DIR)/test-integration.sh $$(go list ./... | grep -v /e2e)
 
 # TODO(user): To use a different vendor for e2e tests, modify the setup under 'tests/e2e'.
 # The default setup assumes Kind is pre-installed and builds/loads the Manager Docker image locally.
@@ -242,7 +242,6 @@ $(LOCALBIN):
 export PATH := $(abspath $(LOCALBIN)):$(PATH)
 
 ## Tool Binaries
-ENVTEST ?= $(LOCALBIN)/setup-envtest
 CLUSTERCTL ?= $(LOCALBIN)/clusterctl
 GARDENER ?= $(LOCALBIN)/gardener
 CAPI ?= $(LOCALBIN)/capi
@@ -250,33 +249,12 @@ APIGEN ?= $(LOCALBIN)/apigen
 KCP ?= $(LOCALBIN)/kcp
 
 ## Tool Versions
-# renovate: datasource=github-releases depName=kubernetes-sigs/controller-runtime
-ENVTEST_VERSION ?= v0.21.0
-# renovate: datasource=github-tags depName=kubernetes/api
-ENVTEST_K8S_VERSION ?= v0.33.2
 # renovate: datasource=github-releases depName=kubernetes-sigs/cluster-api
 CLUSTERCTL_VERSION ?= v1.10.3
 # renovate: datasource=github-releases depName=kcp-dev/kcp
 APIGEN_VERSION ?= v0.27.1
 # renovate: datasource=github-releases depName=kcp-dev/kcp
 KCP_VERSION ?= v0.27.1
-
-# Build ENVTEST release branch names to fetch the envtest setup script (i.e. release-0.20)
-ENVTEST_BRANCH_NAME ?= $(shell echo $(ENVTEST_VERSION) | awk -F'[v.]' '{printf "release-%d.%d", $$2, $$3}')
-ENVTEST_K8S_BRANCH_NAME ?= $(shell echo $(ENVTEST_K8S_VERSION) | awk -F'[v.]' '{printf "1.%d", $$3}')
-
-.PHONY: setup-envtest
-setup-envtest: envtest ## Download the binaries required for ENVTEST in the local bin directory.
-	@echo "Setting up envtest binaries for Kubernetes version $(ENVTEST_K8S_BRANCH_NAME)..."
-	@$(ENVTEST) use $(ENVTEST_K8S_BRANCH_NAME) --bin-dir $(LOCALBIN) -p path || { \
-		echo "Error: Failed to set up envtest binaries for version $(ENVTEST_K8S_BRANCH_NAME)."; \
-		exit 1; \
-	}
-
-.PHONY: envtest
-envtest: $(ENVTEST) ## Download setup-envtest locally if necessary.
-$(ENVTEST): $(LOCALBIN)
-	$(call go-install-tool,$(ENVTEST),sigs.k8s.io/controller-runtime/tools/setup-envtest,$(ENVTEST_BRANCH_NAME))
 
 .PHONY: envsubst
 envsubst:

--- a/internal/controller/controlplane/gardenershootcontrolplane_controller_test.go
+++ b/internal/controller/controlplane/gardenershootcontrolplane_controller_test.go
@@ -6,12 +6,14 @@ package controller
 
 import (
 	"context"
+	"time"
 
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	mcmanager "sigs.k8s.io/multicluster-runtime/pkg/manager"
 	mcreconcile "sigs.k8s.io/multicluster-runtime/pkg/reconcile"
@@ -32,9 +34,14 @@ var _ = Describe("GardenerShootControlPlane Controller", func() {
 		gardenershootcontrolplane := &controlplanev1alpha1.GardenerShootControlPlane{}
 
 		BeforeEach(func() {
-			By("creating the custom resource for the Kind GardenerShootControlPlane")
-			err := k8sClient.Get(ctx, typeNamespacedName, gardenershootcontrolplane)
+			By("Ensuring the custom resource for the Kind GardenerShootControlPlane")
+			var err error
+			Eventually(func() error {
+				err = k8sClient.Get(ctx, typeNamespacedName, gardenershootcontrolplane)
+				return client.IgnoreNotFound(err)
+			}).WithTimeout(10 * time.Second).Should(Succeed())
 			if err != nil && errors.IsNotFound(err) {
+				By("Creating the custom resource for the Kind GardenerShootControlPlane")
 				resource := &controlplanev1alpha1.GardenerShootControlPlane{
 					ObjectMeta: metav1.ObjectMeta{
 						Name:      resourceName,
@@ -44,7 +51,7 @@ var _ = Describe("GardenerShootControlPlane Controller", func() {
 				}
 				Expect(k8sClient.Create(ctx, resource)).To(Succeed())
 			}
-			Eventually(ctx, func() error { return k8sClient.Get(ctx, typeNamespacedName, gardenershootcontrolplane) }).Should(Succeed())
+			Eventually(ctx, func() error { return k8sClient.Get(ctx, typeNamespacedName, gardenershootcontrolplane) }).WithTimeout(10 * time.Second).Should(Succeed())
 		})
 
 		AfterEach(func() {

--- a/internal/controller/controlplane/suite_test.go
+++ b/internal/controller/controlplane/suite_test.go
@@ -47,7 +47,7 @@ func TestControllers(t *testing.T) {
 var _ = BeforeSuite(func() {
 	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
 
-	ctx, cancel = context.WithCancel(context.TODO())
+	ctx, cancel = context.WithCancel(context.Background())
 
 	var err error
 	err = controlplanev1alpha1.AddToScheme(scheme.Scheme)

--- a/internal/controller/infrastructure/gardenerworkerpool_controller_test.go
+++ b/internal/controller/infrastructure/gardenerworkerpool_controller_test.go
@@ -6,12 +6,14 @@ package infrastructure
 
 import (
 	"context"
+	"time"
 
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	mcmanager "sigs.k8s.io/multicluster-runtime/pkg/manager"
 	mcreconcile "sigs.k8s.io/multicluster-runtime/pkg/reconcile"
@@ -32,9 +34,14 @@ var _ = Describe("GardenerWorkerPool Controller", func() {
 		gardenerworkerpool := &infrastructurev1alpha1.GardenerWorkerPool{}
 
 		BeforeEach(func() {
-			By("creating the custom resource for the Kind GardenerWorkerPool")
-			err := k8sClient.Get(ctx, typeNamespacedName, gardenerworkerpool)
+			By("Ensuring the custom resource for the Kind GardenerWorkerPool")
+			var err error
+			Eventually(func() error {
+				err = k8sClient.Get(ctx, typeNamespacedName, gardenerworkerpool)
+				return client.IgnoreNotFound(err)
+			}).WithTimeout(10 * time.Second).Should(Succeed())
 			if err != nil && errors.IsNotFound(err) {
+				By("Creating the custom resource for the Kind GardenerWorkerPool")
 				resource := &infrastructurev1alpha1.GardenerWorkerPool{
 					ObjectMeta: metav1.ObjectMeta{
 						Name:      resourceName,
@@ -44,7 +51,7 @@ var _ = Describe("GardenerWorkerPool Controller", func() {
 				}
 				Expect(k8sClient.Create(ctx, resource)).To(Succeed())
 			}
-			Eventually(ctx, func() error { return k8sClient.Get(ctx, typeNamespacedName, gardenerworkerpool) }).Should(Succeed())
+			Eventually(ctx, func() error { return k8sClient.Get(ctx, typeNamespacedName, gardenerworkerpool) }).WithTimeout(10 * time.Second).Should(Succeed())
 		})
 
 		AfterEach(func() {

--- a/internal/controller/infrastructure/suite_test.go
+++ b/internal/controller/infrastructure/suite_test.go
@@ -47,7 +47,7 @@ func TestControllers(t *testing.T) {
 var _ = BeforeSuite(func() {
 	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
 
-	ctx, cancel = context.WithCancel(context.TODO())
+	ctx, cancel = context.WithCancel(context.Background())
 
 	var err error
 	err = infrastructurev1alpha1.AddToScheme(scheme.Scheme)

--- a/internal/webhook/controlplane/v1alpha1/webhook_suite_test.go
+++ b/internal/webhook/controlplane/v1alpha1/webhook_suite_test.go
@@ -49,7 +49,7 @@ func TestAPIs(t *testing.T) {
 var _ = BeforeSuite(func() {
 	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
 
-	ctx, cancel = context.WithCancel(context.TODO())
+	ctx, cancel = context.WithCancel(context.Background())
 
 	var err error
 	err = controlplanev1alpha1.AddToScheme(scheme.Scheme)

--- a/internal/webhook/infrastructure/v1alpha1/webhook_suite_test.go
+++ b/internal/webhook/infrastructure/v1alpha1/webhook_suite_test.go
@@ -49,7 +49,7 @@ func TestAPIs(t *testing.T) {
 var _ = BeforeSuite(func() {
 	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
 
-	ctx, cancel = context.WithCancel(context.TODO())
+	ctx, cancel = context.WithCancel(context.Background())
 
 	var err error
 	err = infrastructurev1alpha1.AddToScheme(scheme.Scheme)

